src/java.base/share/classes/java/net/DatagramSocket.java

```diff
@@ -1398,14 +1399,10 @@ private static SocketException toSocketException(IOException e) {
         boolean initialized = false;
         try {
             DatagramSocketImplFactory factory = DatagramSocket.factory;
-            if (USE_PLAINDATAGRAMSOCKET || factory != null) {
+            if (factory != null) {
                 // create legacy DatagramSocket delegate
-                DatagramSocketImpl impl;
-                if (factory != null) {
-                    impl = factory.createDatagramSocketImpl();
-                } else {
-                    impl = DefaultDatagramSocketImplFactory.createDatagramSocketImpl(multicast);
-                }
+                DatagramSocketImpl impl = factory.createDatagramSocketImpl();
+                Objects.requireNonNull(impl);
```

Comment:
I assume this Objects.requireNonNull is not needed or maybe you want to detect a buggy factory rather than in the the constructor?

Reply:
Yes - the constructor would throw NPE - but there's no need to call this constructor if `impl` is null here - it's better to fail early.

Reply:
This is probably a case where the spec needs say that an error (maybe an unspecified error) is thrown if the custom factory's createDatagramSocketImpl returns null. We can create a separate issue for that as it's probably existed forever.

Reply:
@AlanBateman 

`protected DatagramSocket(DatagramSocketImpl  impl)` throws NPE if `impl` is `null`.
(this is covered by the blanket statement for NPE)

Do we really need to specify anything else since the global `setDatagramSocketImplFactory` method has been deprecated?

Reply:
Creating an issue to update DatagramSocketImplFactory.createDatagramSocketImpl to say that returning null leads to undefined behavior or DatagramSocket constructors throwing an error is fine, I wasn't suggesting we have to fix it here.

Reply:
I've created an issue to track this: https://bugs.openjdk.java.net/browse/JDK-8269288

Reply:
> I've created an issue to track this: https://bugs.openjdk.java.net/browse/JDK-8269288

Thanks. So are you keeping the Objects.requireNonNull here? If so then it should probably be the 2-arg version so that the message is clear that the custom DatagramSocketFactory returned null, otherwise it will look like a DS bug.

Reply:
@AlanBateman Ah - good idea - something like "implementation returned by installed DatagramSocketFactoryImpl is null"?
Maybe we could add the name of the concrete DatagramSocketFactoryImpl class too?

Reply:
OK. I've added a message to the null check now. See 36bcee3